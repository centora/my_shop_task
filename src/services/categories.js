import { rest } from './rest';

const getInfo = () => rest.get('shop_info');
const getCategories = () => rest.get('public/categories');
const getCategory = id => rest.get(`public/categories/${id}`);
const getProducts = id => rest.get('public/products');
const getProduct = id => rest.get(`public/products/${id}`);

export {
  getInfo,
  getCategories,
  getCategory,
  getProducts,
  getProduct
};
