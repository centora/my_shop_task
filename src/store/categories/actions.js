// 1) get info
// 2) get categories
// 3) get category
// 4) create category
// 5) update category

export const GET_INFO = 'Get info';
export const getInfo = () => ({ type: GET_INFO });

export const SET_INFO = 'Set info';
export const setInfo = data => ({ type: SET_INFO, data });

export const CLEAN_INFO = 'Clean category';
export const cleanInfo = () => ({ type: CLEAN_INFO });

export const GET_CATEGORIES = 'Get Categories';
export const getCategories = () => ({ type: GET_CATEGORIES });

export const UPDATE_CATEGORIES = 'Update Categories';
export const updateCategories = (data) => ({ type: UPDATE_CATEGORIES, data });

export const SET_CATEGORIES = 'Set Categories';
export const setCategories = data => ({ type: SET_CATEGORIES, data });

export const GET_CATEGORY = 'Get Category';
export const getCategory = data => ({ type: GET_CATEGORY, data });

export const SET_CATEGORY = 'Set Category';
export const setCategory = data => ({ type: SET_CATEGORY, data });


