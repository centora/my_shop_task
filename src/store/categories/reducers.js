import {
  SET_INFO,
  CLEAN_INFO,
  GET_CATEGORIES,
  SET_CATEGORIES,
} from './actions';
import {
  info as infoInitial,
  categories as categoriesInitial
} from '../initialState';

export const categories = (state = categoriesInitial, action) => {
  switch (action.type) {
    case GET_CATEGORIES: {
      return state;
    }
    case SET_CATEGORIES: {
      return action.data;
    }
  }

  return state;
};

export const info = (state = infoInitial, action) => {
  switch (action.type) {
    case SET_INFO: {
      return action.data;
    }
    case CLEAN_INFO: {
      return infoInitial;
    }
  }

  return state;
};
