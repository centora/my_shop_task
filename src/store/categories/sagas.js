import {
  takeEvery,
  put,
  all
} from 'redux-saga/effects';

import {
  getInfo,
  updateCategory as updateCategoryService,
  getCategories as getCategoriesService
} from 'services';

import {
  GET_INFO,
  setInfo,
  UPDATE_CATEGORIES,
  GET_CATEGORIES,
  setCategories,
  getCategories,
} from './actions';

function* info() {
  try {
    const info = yield getInfo();
    yield put(setInfo(info));

  } catch (err) {
    console.log(err)
  }
}

function* fetchCategories() {
  try {
    const categories = yield getCategoriesService();
    yield put(setCategories(categories));
  } catch (err) {
    console.log(err);
  }
}

function* putCategory({ data }) {
  try {
    yield updateCategoryService(data);
    yield put(getCategories())
  } catch (err) {
    console.log(err);
  }
}

export function* watchCategories() {
  yield all([
    takeEvery(GET_INFO, info),
    takeEvery(GET_CATEGORIES, fetchCategories),
    takeEvery(UPDATE_CATEGORIES, putCategory),
  ]);
}
