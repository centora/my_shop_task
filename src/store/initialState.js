export const user = null;
export const info = null;
export const category = null;
export const products = [];
export const error = '';
