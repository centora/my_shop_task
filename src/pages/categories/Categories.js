import { connect } from 'react-redux';
import { CategoriesItems } from 'components/categoriesItems';
import { getCategories } from '../../store/category';

const isPublished = category => category.published;
const notPublished = category => !category.published;

class Categories extends Component {
  componentDidMount() {
    this.props.dispatch(getCategories());
  }

  updateCategories = (title, id) => {
    const category = this.props.categories.find(category => category.id === id);
    category.title = title;

    this.props.dispatch(updateCategories(category));
  }

  render() {
    const { categories } = this.props;
    return (
      <section>
        <h1>Categories</h1>
        <CategoriesItems
          publishedItems={categories.filter(isPublished)}
          unpublishedItems={categories.filter(notPublished)}
          onChangeLeftItem={this.updateCategories}
        />

        {/*<br />
        <br />

        <div className="published-categories">
          <h1 className="main-title">Categories</h1>
          <ul className="categories-list">
            {
              categories.map(item => <li key={item.id}><Link to={`categories/${item.id}`}>{item.title}</Link></li>)
            }
          </ul>
        </div>*/}
      </section>
    );
  }
}

const mapState = state => ({
  categories: state.categories
});

export default connect(mapState)(Categories);
