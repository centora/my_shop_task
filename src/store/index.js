import {
  createStore,
  combineReducers,
  applyMiddleware,
  compose
} from 'redux';

import createSagaMiddleware from 'redux-saga';
import { rootSaga } from './rootSaga';
import { user } from './user';
import { error } from './status';
import { info, categories } from './categories';

const rootReducers = combineReducers({
  user,
  info,
  error,
  categories
});


// eslint-disable-next-line
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const sagaMiddleware = createSagaMiddleware();

export const store = createStore(
  rootReducers,
  composeEnhancers(applyMiddleware(sagaMiddleware))
);

export const runSaga = () => sagaMiddleware.run(rootSaga);

