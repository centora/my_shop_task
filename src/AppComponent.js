import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { ToasterContainer } from 'react-toastr';
import { check, logout } from './store/user';
import { cleanInfo, getInfo } from './store/categories';

import { Header } from './components/header';
import { Footer } from './components/footer';
import { Main } from './components/main';

import { Pages } from './pages/Pages';
import { Modal } from 'components/modal';

class AppComponent extends Component {

  state = {
    showModal: false
  }

  componentDidMount() {
    this.props.dispatch(check());
  }

  componentDidUpdate(prevProps) {
    if (!prevProps.user && this.props.user) {
      this.props.dispatch(getInfo());
    }

    if (prevProps.user && !this.props.user) {
      this.props.history.push('/');
    }
  }

  onLogout = () => {
    this.props.dispatch(logout());
    this.props.dispatch(cleanInfo());
  }

  onOpenModal = () => {
    this.setState({showModal: true})
  }

  onHideModal = () => {
    this.setState({showModal: false})
  }

  onOk = () => {
    console.log('removing ...');
    this.setState({showModal: false})
  }

  render() {
    const { user, info, products } = this.props;
    return (
      <>
        <Header
          user={user}
          info={info}
          onLogout={this.onLogout}
        />
        <Main>
          <Pages
            user={user}
            info={info}
            products={products}
          />
        </Main>
        <button onClick={this.onOpenModal}>Open Modal</button>
        <Footer />

        {/*<ToastContainer
          ref={ref => this.container = ref}
          className="toast-top-right"
        />*/}
        <Modal
          isOpen={this.state.showModal}
          close={this.onHideModal}
          success={this.onOk}
          text="Some warning was happend when you tried to delete category"
        />
      </>
    );
  }
}

const mapState = state => ({
  user: state.user,
  info: state.info,
  error: state.error,
  products: state.products,
  category: state.category,
});

export default withRouter(connect(mapState)(AppComponent));
