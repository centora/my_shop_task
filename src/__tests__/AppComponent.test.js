import { shallow } from 'enzyme';
import { AppComponent } from '../AppComponent';
import { getInfo, cleanInfo } from '../store/categories'


describe('<AppComponent/>', () => {
  it('should match to snapshot', () => {
    const wrapper = shallow(<AppComponent dispatch={ _ => _}/>);
    expect(wrapper).toMatchSnapshot();
  })
  it('should render Header and Main', () => {
  const wrapper = shallow(<AppComponent dispatch={ _ => _}/>);
    expect(wrapper.find('Header').length).toBe(1)
    expect(wrapper.find('Main').length).toBe(1)
  })
  it('should send action get info if user changed', () => {
    const fakeDispatch = jest.fn();
    const wrapper = shallow(<AppComponent dispatch={fakeDispatch} />);
    const user = {};
    wrapper.setProps({  user });
    expect(fakeDispatch).toHaveBeenCalledWith(getInfo());
  })

  it('should send push to history if user changed', () => {
    const fakeDispatch = jest.fn();
    const history = { push: jest.fn() }
    const user = {};
    const wrapper = shallow(<AppComponent history={history} user={user} dispatch={fakeDispatch} />);

    wrapper.setProps({ user: null });
    expect(history.push).toHaveBeenCalledWith('/');
  })

})
