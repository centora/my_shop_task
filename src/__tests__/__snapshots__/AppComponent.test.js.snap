// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppComponent/> should match to snapshot 1`] = `
<Fragment>
  <Header
    onLogout={[Function]}
  />
  <Main>
    <Pages />
  </Main>
  <button
    onClick={[Function]}
  >
    Open Modal
  </button>
  <Footer />
  <Modal
    close={[Function]}
    isOpen={false}
    success={[Function]}
    text="Some warning was happend when you tried to delete category"
  />
</Fragment>
`;
